docs/proposal/20210311-single-controller-multitenancy.md

```diff
@@ -0,0 +1,317 @@
+
+* Referencing a cluster scoped `VSphereClusterIdentity` to be used for credentials.
+* Referencing a `Secret` that contains the credentials in the same namespace as the cluster.
+
+Changed Resources
+
+* `VSphereCluster`
+
+New Resources
+
+* A cluster-scoped `VSphereClusterIdentity` represents the vCenter account to use for reconciliation. This type should also contain references to namespaces that are allowed to use the account.
+
+Changes to VSphereCluster
+A new field is added to the VSphereClusterSpec to reference the `VSphereClusterIdentity`. We intend to use a `VSphereIdentityReference` type, similar to that of `corev1.TypedLocalObjectReference` in order to ensure the only objects that can be referenced are either in the same namespace or are scoped globally.
+
+```go
+// VSphereClusterIdentity is the account to be used for vcenter actions
+type VSphereClusterIdentity struct {
+    metav1.TypeMeta `json:`",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec VSphereClusterIdentitySpec `json:"spec,omitempty"`
+    Status VSphereClusterIdentityStatus `json:"status, omitempty"`
+}
+
+type VSphereClusterIdentitySpec struct {
+    // secretsRef references a secret in the CAPV controller namespace with the credentials to use
+    secretRef `json:"secretRef,omitempty"`
+
+    // AllowedNamespaces is used to identify which namespaces are allowed to use this account.
+    // Namespaces can be selected either by using a label selector.
+    // If this object is nil, no namespaces will be allowed.
+    // If this object is empty {}, clusters can use this identity from any namespace.
+    //
+    // +optional
+    AllowedNamespaces *AllowedNamespaces `json:"allowedNamespaces"`
+}
+
+type AllowedNamespaces struct {
+    // Selector is a standard Kubernetes LabelSelector.
+    // This is a standard Kubernetes LabelSelector, a label query over a set of resources.
+    // The result of matchLabels and matchExpressions are ANDed.
+    // +optional
+    Selector metav1.LabelSelector `json:"selector"`
+}
+
+type VSphereIdentityReference struct {
+    // Kind of the identity.
+    // +kubebuilder:validation:Enum=VSphereClusterIdentity,Secret
+    Kind string
+
+    // Name of the identity
+    // +kubebuilder:validation:MinLength=1
+    Name string
+}
+
+type VSphereClusterSpec struct {
+    ...
+    // +optional
+    IdentityRef *VSphereIdentityReference `json:"IdentityRef,omitempty"`
+}
+```
+
+#### Controller Changes
+
+* If IdentityRef is specified, and it references a `VSphereClusterIdentity`, the CRD is fetched and used to create a session
+* The secret referenced within `VSphereClusterIdentity` must be in the controller namespace
+* If IdentityRef is specified, and it references a `Secret`, the `Secret` is used to create a session. The secret must be in the same namespace as the `VSphereCluster`
```

Comment:
> The secret must be in the same namespace as the `VSphereCluster`

Am I wrong or the secrets for Global resources should be in controller namespace?

Reply:
this is for the local credentials (i.e. BYO creds within the namespace where you provision your cluster). But I agree for the global ones

Reply:
yeah when the secret is addressed through the global `VSphereClusterIdentity` then it's expected to be in the controller namespace. CAPV is also implementing a direct reference to a secret from a `VSphereCluster` and for that we expect it to be in the same namespace as the cluster.